File: packages/protocol/src/operations.ts
/**
 * ScreenMesh syncs by exchanging operations, not whole workspaces.
 * Devices exchange only the operations they are missing.
 * See docs/Architecture.md §2 (Layer 3).
 */

export type OperationType =
  | "CREATE_OBJECT"
  | "UPDATE_OBJECT"
  | "DELETE_OBJECT"
  | "SEND_TO_DEVICE"
  | "MARK_DELIVERED"
  | "MARK_OPENED"
  | "PIN_OBJECT"
  | "MOVE_OBJECT"
  | "ADD_ATTACHMENT"
  | "REVOKE_DEVICE"
  | "YJS_UPDATE"
  | "CONTINUE_ON_DEVICE"
  | "CARRY_BUNDLE"
  | "REJECT_OBJECT"
  | "FILE_CHUNK"
  | "FILE_CHUNK_ACK"
  | "EDIT_PRESENCE";

export interface Operation<TPayload = unknown> {
  operationId: string;
  deviceId: string;
  workspaceId: string;
  type: OperationType;
  objectId?: string;
  timestamp: number;
  payload: TPayload;
}

/** A device's view of how far it has seen each peer's oplog. */
export type SyncVector = Record<string, number>;

import type { MeshObject, MeshObjectType, SendOptions } from "./types.js";

export interface CreateObjectPayload {
  object: MeshObject;
}

export interface SendToDevicePayload {
  objectId: string;
  options?: SendOptions;
}

export interface ObjectRefPayload {
  objectId: string;
}

export interface RevokeDevicePayload {
  deviceId: string;
}

/** A Yjs document update for collaborative editing of an object. */
export interface YjsUpdatePayload {
  objectId: string;
  updateB64: string;
}

/** Last-write-wins content replacement (checklist toggles, etc.). */
export interface UpdateObjectPayload {
  objectId: string;
  content: unknown;
  updatedAt: number;
  /** Present when a reclassification changes the object's type (e.g. a
   *  misdetected checklist corrected to a document) alongside its content. */
  type?: MeshObjectType;
}

/** Ask the target device to open this object for editing. */
export interface ContinueOnDevicePayload {
  objectId: string;
}

/**
 * Store–carry–forward (docs/Architecture.md §2): hand an opaque, already
 * end-to-end-encrypted envelope to a peer for safekeeping, addressed to a
 * third device the sender can't currently reach. Since envelope payloads
 * are now encrypted with a per-pair ratchet key (see
 * packages/crypto/src/ratchet.ts), the carrier genuinely cannot decrypt
 * the inner envelope — it doesn't hold the ratchet session between the
 * original sender and the true destination, unlike the relay/carrier
 * under the old shared-workspace-key model. `encryptedPayloadB64` is the
 * base64 JSON of an EnvelopeJson.
 */
export interface CarryBundlePayload {
  bundleId: string;
  sourceDeviceId: string;
  destinationDeviceId: string;
  encryptedPayloadB64: string;
  createdAt: number;
  expiresAt: number;
  /** Remaining number of DISTINCT carriers this bundle may still be handed to. */
  hopLimit: number;
}

/**
 * Secure file drop (docs/Roadmap.md Phase 5): larger files are split into
 * chunks, each sent as its own encrypted envelope (own ratchet message,
 * own carry-eligibility) rather than one giant envelope. The receiver
 * materializes the MeshObject only once every chunk has arrived; `meta`
 * carries everything needed to do that and is only sent on chunk 0 to
 * avoid repeating it in every chunk.
 */
export interface FileChunkMeta {
  objectType: "file" | "image";
  name: string;
  mimeType: string;
  size: number;
  createdBy: string;
  createdAt: number;
  expiresAt?: number;
  options?: SendOptions;
}

export interface FileChunkPayload {
  fileId: string;
  chunkIndex: number;
  totalChunks: number;
  dataB64: string;
  meta?: FileChunkMeta;
}

/**
 * Sent by the receiver immediately after it durably persists a chunk (see
 * ScreenMeshDb's `fileChunks` table), so the sender can tell which chunks
 * actually landed and resend only the ones that didn't — instead of either
 * trusting a fire-and-forget send or resending the whole file blind.
 */
export interface FileChunkAckPayload {
  fileId: string;
  chunkIndex: number;
}

/**
 * Ephemeral "someone is editing this" signal — a UX safeguard, not part of
 * the durable object model. Sent as a heartbeat while a device has an
 * editable object's editor open, and once more with `active: false` on
 * close. Never blocks another device from also editing; it exists only so
 * a person can see they might be about to collide with someone else,
 * since the underlying Yjs merge can't fully protect a peer's first edit
 * on an object it never got seeded Yjs state for (see engine.ts's
 * editText doc comment).
 */
export interface EditPresencePayload {
  objectId: string;
  active: boolean;
}

File: apps/android/app/src/main/java/com/screenmesh/protocol/Operations.kt
package com.screenmesh.protocol

import kotlinx.serialization.Serializable
import kotlinx.serialization.json.JsonElement

/**
 * Kotlin mirror of packages/protocol/src/operations.ts. `payload` stays a
 * raw JsonElement (not a sealed/polymorphic type) — exactly like the TS
 * side, which types it `unknown` and casts per `op.type` at the call
 * site (`op.payload as SendToDevicePayload`). This keeps the wire format
 * identical without needing kotlinx.serialization's polymorphic
 * discriminator config to also match the TS shape exactly.
 */
object OperationTypes {
    const val CREATE_OBJECT = "CREATE_OBJECT"
    const val UPDATE_OBJECT = "UPDATE_OBJECT"
    const val DELETE_OBJECT = "DELETE_OBJECT"
    const val SEND_TO_DEVICE = "SEND_TO_DEVICE"
    const val MARK_DELIVERED = "MARK_DELIVERED"
    const val MARK_OPENED = "MARK_OPENED"
    const val PIN_OBJECT = "PIN_OBJECT"
    const val MOVE_OBJECT = "MOVE_OBJECT"
    const val ADD_ATTACHMENT = "ADD_ATTACHMENT"
    const val REVOKE_DEVICE = "REVOKE_DEVICE"
    const val YJS_UPDATE = "YJS_UPDATE"
    const val CONTINUE_ON_DEVICE = "CONTINUE_ON_DEVICE"
    const val CARRY_BUNDLE = "CARRY_BUNDLE"
    const val REJECT_OBJECT = "REJECT_OBJECT"
    const val FILE_CHUNK = "FILE_CHUNK"
}

@Serializable
data class Operation(
    val operationId: String,
    val deviceId: String,
    val workspaceId: String,
    val type: String,
    val objectId: String? = null,
    val timestamp: Long,
    val payload: JsonElement,
)

@Serializable
data class CreateObjectPayload(val `object`: MeshObject)

@Serializable
data class SendToDevicePayload(val objectId: String, val options: SendOptions? = null)

@Serializable
data class ObjectRefPayload(val objectId: String)

@Serializable
data class RevokeDevicePayload(val deviceId: String)

/** A Yjs document update — NOT yet applied by MeshEngine.kt (no Yjs-for-Kotlin port yet). */
@Serializable
data class YjsUpdatePayload(val objectId: String, val updateB64: String)

/** Last-write-wins content replacement (checklist toggles, etc.). */
@Serializable
data class UpdateObjectPayload(val objectId: String, val content: JsonElement, val updatedAt: Long)

/** Applied by MeshEngine.kt's continueOnDevice / CONTINUE_ON_DEVICE case. */
@Serializable
data class ContinueOnDevicePayload(val objectId: String)

/** Store–carry–forward hand-off — applied by MeshEngine.kt's CARRY_BUNDLE case. */
@Serializable
data class CarryBundlePayload(
    val bundleId: String,
    val sourceDeviceId: String,
    val destinationDeviceId: String,
    val encryptedPayloadB64: String,
    val createdAt: Long,
    val expiresAt: Long,
    val hopLimit: Int,
)

/** Secure file drop chunk — applied by MeshEngine.kt's sendFileChunks / FILE_CHUNK case. */
@Serializable
data class FileChunkMeta(
    val objectType: String,
    val name: String,
    val mimeType: String,
    val size: Long,
    val createdBy: String,
    val createdAt: Long,
    val expiresAt: Long? = null,
    val options: SendOptions? = null,
)

@Serializable
data class FileChunkPayload(
    val fileId: String,
    val chunkIndex: Int,
    val totalChunks: Int,
    val dataB64: String,
    val meta: FileChunkMeta? = null,
)
